### ISSUE TITLE
Bogus error codes from File::Copy::move

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6560]
Migrated from [rt.perl.org#22622](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=22622) (status was 'resolved')

Searchable as RT22622$

**Comment by p5pRT** [comment]
### From Eric.Amick@mail.house.gov

#### Created by eric.amick@mail.house.gov



The following script demonstrates the problem\.  The directory testdir is
not writable by the user\, and testdir and /tmp are on different
filesystems\.\.\.

use File&#8203;::Copy \<File&#8203;::Copy> ;
move\("testdir/testfile"\, "/tmp"\) or die "$\!\\n";

The error message is "Cross\-device link"\, which is true but irrelevant\.
Since move attempts to rename first\, then tries copy and unlink if the
rename fails\, the error message should reflect not the rename call 
failure but the copy/unlink process failure\.  The routine is saving the
error codes from the rename call and restoring them before returning\,
but it should do that with the copy/unlink error codes instead\.  The
following patch will fix the problem&#8203;:

\*\*\* /opt/perl/lib/5\.6\.1/File/Copy\.pm    Thu Jun 21 00&#8203;:55&#8203;:51 2001
\-\-\- File/Copy\.pm    Mon Jun  9 09&#8203;:46&#8203;:07 2003
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 147\,153 \*\*\*\*
&nbsp;  
&nbsp;  sub move \{
&nbsp;      my\($from\,$to\) = @&#8203;\_;
\!     my\($copied\,$fromsz\,$tosz1\,$tomt1\,$tosz2\,$tomt2\,$sts\,$ossts\);
&nbsp;  
&nbsp;      if \(\-d $to && \! \-d $from\) \{
&nbsp;      $to = \_catname\($from\, $to\);
\-\-\- 147\,153 \-\-\-\-
&nbsp;  
&nbsp;  sub move \{
&nbsp;      my\($from\,$to\) = @&#8203;\_;
\!     my\($fromsz\,$tosz1\,$tomt1\,$tosz2\,$tomt2\,$sts\,$ossts\);
&nbsp;  
&nbsp;      if \(\-d $to && \! \-d $from\) \{
&nbsp;      $to = \_catname\($from\, $to\);
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 161\,167 \*\*\*\*
&nbsp;      \}
&nbsp;      return 1 if rename $from\, $to;
&nbsp;  
\-     \($sts\,$ossts\) = \($\! \+ 0\, $^E \+ 0\);
&nbsp;      \# Did rename return an error even though it succeeded\, because $to
&nbsp;      \# is on a remote NFS file system\, and NFS lost the server's ack?
&nbsp;      return 1 if defined\($fromsz\) && \!\-e $from &&           \# $from
disappeared
\-\-\- 161\,166 \-\-\-\-
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
\*\*\* 170\,177 \*\*\*\*
&nbsp;                  $tosz2 == $fromsz;                         \# it's all
there
&nbsp;  
&nbsp;      \($tosz1\,$tomt1\) = \(stat\($to\)\)\[7\,9\];  \# just in case rename did
something
\!     return 1 if \($copied = copy\($from\,$to\)\) && unlink\($from\);
&nbsp;  
&nbsp;      \($tosz2\,$tomt2\) = \(\(stat\($to\)\)\[7\,9\]\,0\,0\) if defined $tomt1;
&nbsp;      unlink\($to\) if \!defined\($tomt1\) or $tomt1 \!= $tomt2 or $tosz1 \!=
$tosz2;
&nbsp;      \($\!\,$^E\) = \($sts\,$ossts\);
\-\-\- 169\,177 \-\-\-\-
&nbsp;                  $tosz2 == $fromsz;                         \# it's all
there
&nbsp;  
&nbsp;      \($tosz1\,$tomt1\) = \(stat\($to\)\)\[7\,9\];  \# just in case rename did
something
\!     return 1 if copy\($from\,$to\) && unlink\($from\);
&nbsp;  
\+     \($sts\,$ossts\) = \($\! \+ 0\, $^E \+ 0\);
&nbsp;      \($tosz2\,$tomt2\) = \(\(stat\($to\)\)\[7\,9\]\,0\,0\) if defined $tomt1;
&nbsp;      unlink\($to\) if \!defined\($tomt1\) or $tomt1 \!= $tomt2 or $tosz1 \!=
$tosz2;
&nbsp;      \($\!\,$^E\) = \($sts\,$ossts\);

\(I took the liberty of deleting an unused variable as well\.\)  I tested
this on 5\.6\.1\, but this should work on any version of Perl where this
module can be installed\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.1:

Configured by gsar at Wed Jun 20 21:38:27 PDT 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=hpux, osvers=11.00, archname=PA-RISC1.1-thread-multi
    uname='hp-ux llbertha b.11.00 u 9000800 2002402864 unlimited-user
license '
    config_args='-des -Dcf_by=ActiveState
-Dcf_email=ActivePerl@ActiveState.com -Uinstallusrbinperl -Dusethreads
-Duseithreads -Duselargefiles -Dinc_version_list=5.6.0/$archname 5.6.0
-Dcc=gcc -Accflags=-mpa-risc-1-1 -fPIC -Dcccdlflags=-fPIC -Dd_attribut=undef
-Darchname=PA-RISC1.1 -Dprefix=/opt/perl'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define
usemultiplicity=define
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='gcc', ccflags ='-D_POSIX_C_SOURCE=199506L -D_HPUX_SOURCE
-L/lib/pa1.1 -DUINT32_MAX_BROKEN -mpa-risc-1-1 -fPIC -fno-strict-aliasing
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O',
    cppflags='-D_POSIX_C_SOURCE=199506L -D_HPUX_SOURCE -L/lib/pa1.1
-DUINT32_MAX_BROKEN -mpa-risc-1-1 -fPIC -fno-strict-aliasing
-I/usr/local/include'
    ccversion='', gccversion='2.9-hppa-991112', gccosandvers='hpux11.00'
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='ld', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/ccs/lib
    libs=-lnsl -lnm -lndbm -lmalloc -ldld -lm -lpthread -lc -lndir -lcrypt
-lsec
    perllibs=-lnsl -lnm -lmalloc -ldld -lm -lpthread -lc -lndir -lcrypt
-lsec
    libc=/lib/libc.sl, so=sl, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_hpux.xs, dlext=sl, d_dlsymun=undef, ccdlflags='-Wl,-E '
    cccdlflags='-fPIC', lddlflags='-b -L/usr/local/lib'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY


@INC for perl v5.6.1:
    /opt/perl/lib/5.6.1/PA-RISC1.1-thread-multi
    /opt/perl/lib/5.6.1
    /opt/perl/lib/site_perl/5.6.1/PA-RISC1.1-thread-multi
    /opt/perl/lib/site_perl/5.6.1
    /opt/perl/lib/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/erica
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
 
PATH=/bin:/usr/bin:/usr/dt/bin:/usr/bin/X11:/opt/ansic/bin:/usr/ccs/bin:/opt
/perl/bin:/usr/contrib/bin:/opt/nettladm/bin:/opt/fc/bin:/opt/fcms/bin:/opt/
upgrade/bin:/opt/pd/bin:/usr/contrib/bin/X11:/opt/OpenSource/bin:/opt/graphi
cs/phigs/bin:/opt/graphics/common/bin:/opt/resmon/bin:/usr/sbin/diag/contrib
:/opt/hparray/bin:/opt/langtools/bin:/opt/imake/bin:/opt/hpnpl//bin:/opt/net
scape:/evs/oracle/OraHome1/bin:/home/erica/bin:.
    PERL_BADLANG (unset)
    SHELL=/usr/bin/ksh
    SHLIB_PATH (unset)


Eric Amick
Legislative Computer Systems


```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Amick\, Eric wrote&#8203;:
> The following script demonstrates the problem\.  The directory testdir is
> not writable by the user\, and testdir and /tmp are on different
> filesystems\.\.\.
> 
> use File&#8203;::Copy \<File&#8203;::Copy> ;
> move\("testdir/testfile"\, "/tmp"\) or die "$\!\\n";
> 
> The error message is "Cross\-device link"\, which is true but irrelevant\.
> Since move attempts to rename first\, then tries copy and unlink if the
> rename fails\, the error message should reflect not the rename call 
> failure but the copy/unlink process failure\.  The routine is saving the
> error codes from the rename call and restoring them before returning\,
> but it should do that with the copy/unlink error codes instead\.  The
> following patch will fix the problem&#8203;:
> 
> \*\*\* /opt/perl/lib/5\.6\.1/File/Copy\.pm    Thu Jun 21 00&#8203;:55&#8203;:51 2001
> \-\-\- File/Copy\.pm    Mon Jun  9 09&#8203;:46&#8203;:07 2003

Thanks\, I've applied your patch to the development sources of perl\,
that will lead to 5\.10\.0\. It will eventually be integrated into 5\.8\.1\.
\(Change 19799\)\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'
